numpyro/distributions/util.py

```diff
@@ -13,21 +13,144 @@
 from jax.util import partial
 
 
-@partial(jit, static_argnums=(3, 4))
-def _binomial(key, p, n, n_max, shape):
+_tr_params = namedtuple('tr_params', ['c', 'b', 'a', 'alpha', 'u_r', 'v_r'])
+
+
+def _get_tr_params(n, p):
+    spq = np.sqrt(n * p * (1 - p))
+    c = n * p + 0.5
+    b = 1.15 + 2.53 * spq
+    a = -0.0873 + 0.0248 * b + 0.01 * p
+    alpha = (2.83 + 5.1 / b) * spq
+    u_r = 0.43
+    v_r = 0.92 - 4.2 / b
+    return _tr_params(c, b, a, alpha, u_r, v_r)
+
+
+def stirling_approx_tail(k):
+    precomputed = [
+        0.0810614667953272,
+        0.0413406959554092,
+        0.0276779256849983,
+        0.02079067210376509,
+        0.0166446911898211,
+        0.0138761288230707,
+        0.0118967099458917,
+        0.0104112652619720,
+        0.00925546218271273,
+        0.00833056343336287,
+    ]
+    return np.where(k < 10,
+                    np.take(precomputed, k),
+                    1. / (12 * (k + 1)) - 1./(360 * (k + 1) ** 3) + 1./(1260 * (k + 1) ** 5))
+
+
+def _binomial_btrs(key, p, n):
+    """
+    Based on the transformed rejection sampling algorithm (BTRS) from the
+    following reference:
+
+    Hormann, "The Generation of Binonmial Random Variates"
+    (https://core.ac.uk/download/pdf/11007254.pdf)
+    """
+    def _btrs_body_fn(val):
+        _, key, p, n, tr_params, _, _ = val
+        key, key_u, key_v = random.split(key, 3)
+        u = random.uniform(key_u)
+        v = random.uniform(key_v)
+        u = u - 0.5
+        k = np.floor((2 * tr_params.a / (0.5 - np.abs(u)) + tr_params.b) * u + tr_params.c).astype(n.dtype)
```

Comment:
`0.43` is directly from Table 1 and corresponds to the squeeze op in transformed rejection sampling algorithm (`|u| <= u_r and v <= vr `). We can pass `us` too, but the paper was written for a different era, and I'm not sure if saving this one computation is worth passing additional state.

Reply:
This term is used in Step 3.1 too, so having this additional state might be easier to follow the reference. But recomputing it is also fine: computing `log` terms are much more expensive than those terms.

Reply:
I suppose I have tried to make the algorithm readable as a TRS algorithm (which is much more intuitive to follow), rather than a verbatim translation of steps for BTRS, but your comment indicates that I should at the very least add comments so that its easier for the reader to follow! 😄

Reply:
Oh, I see. I only look at the algorithm to review. :))

Reply:
Re: astype is required for Stirling. Pls ignore my comment!